Add guidelines for single approval PRs (#6785)
sunpy/sunpy: docs/dev_guide/contents/pr_review_procedure.rst

* All the continuous integration must pass unless there is a known issue.

* At least two members (excluding the PR author) of the "sunpy-developers" group must have approved the PR.

>   One should, ideally, be a relevant subpackage maintainer.

Reviewer: I wonder if this line is still relevant outside of coordinates.

Author: Yeah, we got rid of subpackage maintainers right? I could just remove this line?

Reviewer: I think it should go.

* All comments posted on the thread must be resolved.

> It is important that approval for merging the PR is always done on the comment thread, as this becomes part of the "permanent record".

Reviewer: We don't normally comment but approve the PR via the Github UI. Sometimes people add a comment on the approval but I don't see why would need to be required.

Author: Yeah, not sure what this meant originally - will update and clarify.

Reviewer 2: Maybe it meant your approval must be a green tick on GitHub and not just a message on Matrix saying you approve merging?

Author: 👍  updated text here to make it clearer

Minor changes
-------------

> If a PR only makes minor changes, it can be merged by the first reviewer, if they are confident they fully understand the changes.

Reviewer 3: How are we defining "minor changes" here? A bugfix? A documentation rewording? In the past, we've said bugfixes only require a single review. In my mind, a bugfix is more clearly defined than a "minor change." Here, we seem to be broadening the scope, but not defining what that scope is. I don't disagree with the spirit of this, but o avoid just having this same debate on every single PR that someone marks as "minor changes" I think we should make an attempt here to define what a "minor change" actually means.

Reviewer: What metrics could we use? 

I can think of:
- Line count of changes, say less than 10?
- The location of the change, e.,g a CI file vs a Python module
- Others?

Author: I'd be reluctant to have a hard limit on lines. Also reluctant to have hard rules in general, I think we all trust each others judgement when it comes to reviewing? Agree that some guidelines would definitely be helpful though, perhaps one-reviewer PRs could be:

- Improvements to existing documentation (including adding missing docstrings)
- 'Minor' bugfixes (including to the CI), leaving open interpretation of 'minor'
- **Not** new features (including to the CI), no matter how 'minor'
- **Not** new documentation pages (e.g. a new example, tutorial, page)
?

Reviewer: That sounds good to me. Anything new does not count as minor

Reviewer 4: This should simply just be explained by example then, something like:

Exactly what constitutes a minor-change is left up the the reviewer but some examples might include:

- Improvements to existing documentation (including adding missing docstrings)
- 'Minor' bugfixes (including to the CI), leaving open interpretation of 'minor'
- Changes in code style rather than substance

A minor-change can never include:

- New features (including to the CI), no matter how 'minor'
- New documentation pages (e.g. a new example, tutorial, page)
- API changes
- etc.

Reviewer 3: Agree about not having hard rules and trusting each other's judgement. I think enumerating examples of minor changes is sufficient as Alasdair has done above. I would exclude "etc." though and amend API changes to "Changes to the public-facing API". We can always amend this list as needed if at some point we feel it needs more clarification.

Reviewer: I still think the CI should be special cased.

> If this happens, the reviewer who merges should add the ``DocFix`` or ``CodeFix`` labels to the PR to indicate that they took a conscious decision that it only needed on reviewer.

Reviewer: If I recall bugfix also has this one review is required note. Do we want to change that?

Author: I thought that being a bugfix didn't automatically imply the change being a trivial one that only needed one reviewer, so I'd be pro removing that tag (or re-purposing it to denote a bug fix PR).

Reviewer 5: The "Bugfix" label has historically meant that it required only one reviewer

Author: Sorry, perhaps a bit of confusion with my language 😄 - what I meant was that in my opinion fixing a bug should not always imply that a PR only needs one reviewer.

Reviewer 5: No disagreement there

Fixed sentence:
The "Bugfix" label has historically meant that the PR not only fixed a bug, but was also sufficiently simple and localized as to necessitate only one reviewer

Author: I've simplified this sentence now to recommend a single label, `minor-change`, that we can apply if a change is considered minor enough for one reviewer. Thoughts?

> The original PR author should not decide whether the change is minor or not.

Reviewer 5: Hmm, I recognize the concern, but if this is cast as a hard-and-fast rule, I'll probably end up breaking the rule regularly.

Reviewer: I agree with Albert here. The maintainers should be trusted to be label PRs and the reviewer should check if they agree with the choice or not. If they disagree then it should need a second review.

Author: > breaking the rule regularly.

Breaking it by doing the labelling, or breaking it by self-merging once someone else has approved but not merged?

> The maintainers should be trusted to be label PRs and the reviewer should check if they agree with the choice or not. If they disagree then it should need a second review.

👍 I'd be happy to go with this - I'll make a proposed change.

Reviewer 5: >> breaking the rule regularly.
>
> Breaking it by doing the labelling, or breaking it by self-merging once someone else has approved but not merged?

My comment is moot given subsequent changes, but for the record I'll clarify that I meant that I would continue to add labels to my own PRs if I judge them to be minor changes.

* All the continuous integration must pass unless there is a known issue.

> * At least two members (excluding the PR author) of the "sunpy-developers" group must have approved the PR.

Reviewer: This is an aside for this PR, but this would become project members in future?

Reviewer 4: My understanding wws from the SEP is that the two groups (project members and sunpy maintainers) remain distinct, even though all maintainers are to be made members, members wouldn't be made made maintainers necessarily when voted in. This is probably worth raising on the SEP pr for clarification though.

Reviewer 6: yeah, maintainers of packages aren't the same group as project members. Not all project members will be developers, and certainly not developers of core or all packages etc.

  Exceptions can be made for minor changes, see below.

* All comments posted on the thread must be resolved.

It is important that approval for merging the PR is always done by explicitly approving the PR through the GitHub UI before merging, so a record of approval is left in the PR thread.

Minor changes
-------------

If a PR only makes minor changes, it can be merged by the first reviewer, if they are confident they fully understand the changes.

> If this happens, the ``minor-change`` label should be added to the PR to indicate that it has been considered minor enough to need only one reviewer.

Reviewer: SO we will retire the bug,doc,code fix labels for this? I like it.

> The PR author can add one of these labels as a suggestion, but the first reviewer can always request a second reviewer if they think the PR would benefit from it.

Reviewer 6: ```suggestion
The PR author can add one of this label as a suggestion, but the first reviewer can always request a second reviewer if they think the PR would benefit from it.
```
?

Reviewer 5: ```suggestion
The PR author can add this label as a suggestion, but the first reviewer can remove the label as part of their evaluation.
```

> Exactly what constitutes a minor-change is left up the the reviewer but some examples might include:

Reviewer 6: ```suggestion
Exactly what constitutes a minor-change is left up to the the reviewer but some examples might include:
```
